contributors/design-proposals/storage/svcacct-token-volume-source.md

```diff
@@ -0,0 +1,164 @@
+// the pods runtime filesystem for use against APIs (Kubernetes API Server or
+// otherwise).
+type ServiceAccountTokenProjection struct {
+  // Audience is the intended audience of the token. A recipient of a token
+  // must identify itself with an identifier specified in the audience of the
+  // token, and otherwise should reject the token.
+  Audience string
+  // ExpirationSeconds is the requested duration of validity of the service
+  // account token. Defaults to 1 hour.
+  ExpirationSeconds *int64
+  // Path is the relative path of the file to project the token into.
+  Path string
+}
+```
+
+A volume plugin implemented in the kubelet will project a service account token
+sourced from the TokenRequest API into volumes created from
+ProjectedVolumeSources. As the token approaches expiration, the kubelet volume
+plugin will proactively rotate the service account token. The kubelet will start
+trying to rotate the token if the token is older than 80 percent of its time to
+live or if the token is older than 24 hours.
+
+To replace the current service account token secrets, we also need to inject the
+clusters CA certificate bundle. Initially we will deploy to data in a configmap
+per-namespace and reference it using a ConfigMapProjection.
+
+A projected volume source that is equivalent to the current service account
+secret:
+
+```yaml
+sources:
+- serviceAccountToken:
+    expirationSeconds: 3153600000 # 100 years
+    path: token
+- configMap:
+    name: kube-cacrt
+    items:
+    - key: ca.crt
+      path: ca.crt
+- downwardAPI:
+    items:
+    - path: namespace
+      fieldRef: metadata.namespace
+```
+
+
+This fixes one scalability issue with the current service account token
+deployment model where secret GETs are a large portion of overall apiserver
+traffic.
+
+A projected volume source that requests a token for vault and Istio CA:
+
+```yaml
+sources:
+- serviceAccountToken:
+    path: vault-token
+    audience: vault
+- serviceAccountToken:
+    path: istio-token
+    audience: ca.istio.io
+```
+
+### Risks and Mitigations
+
+Reducing the scope of service account tokens by not creating them automatically
+on service account creation is technically an API break. This would have to be
+opt-in, but like RBAC is about reducing the scope of vulnerability. Many people
+may opt not to disable it. For those who do disable it, we can preserve the
+existing behavior of being able to create a secret of type service-account-token
+(with the annotation to the service account that links to it) and have the
+controller auto populate it.
+
+FlexVolume and CSI are the only way to deliver custom content to nodes today
+without persisting it in the API. In a virtual kubelet environment, these
+mechanisms may not work the same as on regular kubelets, so third party identity
+integrators may not be able to deliver their custom content. A container volume
+or init container might be a sufficient workaround.
```

Comment:
I also don't understand this paragraph. Is this an alternative implementation? I thought tokens from the tokenrequest API aren't persisted in the etcd, is that not true?